odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+            of the space (usually real or complex number).
+            See ``Notes`` for mathematical requirements.
+
+            This option cannot be combined with ``weight``,
+            ``dist`` or ``norm``. It also cannot be used in case of
+            non-numeric ``dtype``.
+
+        dist_using_inner : bool, optional
+            Calculate ``dist`` using the formula
+
+                ``||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>``
+
+            This avoids the creation of new arrays and is thus faster
+            for large arrays. On the downside, it will not evaluate to
+            exactly zero for equal (but not identical) ``x`` and ``y``.
+
+            This option can only be used if ``exponent`` is 2.0.
+
+            Default: False.
+
+        kwargs :
+            Further keyword arguments are passed to the weighting
+            classes.
+
+        See Also
+        --------
+        rn : constructor for real tensor spaces
+        cn : constructor for complex tensor spaces
+        tensor_space :
+            constructor for tensor spaces of arbitrary scalar
+            data type
+
+
+        Notes
+        -----
+        - A distance function or metric on a space :math:`\mathcal{X}`
+          is a mapping
+          :math:`d:\mathcal{X} \\times \mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all space elements
+          :math:`x, y, z`:
+
+          * :math:`d(x, y) \geq 0`,
+          * :math:`d(x, y) = 0 \Leftrightarrow x = y`,
+          * :math:`d(x, y) = d(y, x)`,
+          * :math:`d(x, y) \\leq d(x, z) + d(z, y)`.
+
+        - A norm on a space :math:`\mathcal{X}` is a mapping
+          :math:`\| \cdot \|:\mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all
+          space elements :math:`x, y`: and scalars :math:`s`:
+
+          * :math:`\| x\| \geq 0`,
+          * :math:`\| x\| = 0 \Leftrightarrow x = 0`,
+          * :math:`\| sx\| = |s| \cdot \| x \|`,
+          * :math:`\| x+y\| \\leq \| x\| +
+            \| y\|`.
+
+        - An inner product on a space :math:`\mathcal{X}` over a field
+          :math:`\mathbb{F} = \mathbb{R}` or :math:`\mathbb{C}` is a
+          mapping
+          :math:`\\langle\cdot, \cdot\\rangle: \mathcal{X} \\times
+          \mathcal{X} \\to \mathbb{F}`
+          satisfying the following conditions for all
+          space elements :math:`x, y, z`: and scalars :math:`s`:
+
+          * :math:`\\langle x, y\\rangle =
+            \overline{\\langle y, x\\rangle}`,
+          * :math:`\\langle sx + y, z\\rangle = s \\langle x, z\\rangle +
+            \\langle y, z\\rangle`,
+          * :math:`\\langle x, x\\rangle = 0 \Leftrightarrow x = 0`.
+
+        Examples
+        --------
+        >>> space = NumpyTensorSpace((2, 3), dtype='float64')
+        >>> space
+        rn((2, 3))
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='complex64')
```

Comment:
Note earlier comment on shape. Should be integral in general unless having several shows something.